Handle Managed Schemas in the LogTypesAPI

alxarch opened the pull request:
## Background

LogTypes API  updated to support both user-defined and managed log types.

Closes #2440 
Closes #2437 

## Changes

- Add `logtypesapi.SchemaRecord` to reperesent both user-defined and managed schemas.
- Replace `logtypesapi.LogTypesDatabase` interface with `logtypesapi.SchemaDatabase` to handle operations for all schemas.
- Update `CustomLog` endpoints to use `SchemaDatabase` in a backwards compatible way  (e.g. conditions on  `Managed != true`)
- Add `GetSchema` endpoint to return any schema
- Add `ListManagedSchemaUpdates` endpoint to list managed schema updates from `panther-analysis` repo
- Add the `UpdateManagedSchemas` endpoint to load a managed schemas manifest from a URL and update all entries found within.
- Embeds the minimum required release version for managed schemas using `go-bindata`
- Add a `build.sh` to use when we want to update the minimum required release version for managed schemas embedded in panther
- Adds a new custom resource  `Custom::ManagedSchemas` to handle managed schema updates during a deployment


## Testing

- mage test:go
- e2e deployed and checked that entries in DDB are created and that log processor/ data catalog updater can resolve log types properly using the logtypes API

alxarch commented:
This is a long PR. The changes to LogTypesAPI are extensive because we need to keep the current endpoints that were designed with only custom logs in mind. In order to manage both user-defined and managed logs in the same db, we introduce the `SchemaRecord` type that represents both user-defined and panther-managed schemas.

The `LogTypesAPI` is pivoting towards `SchemasAPI`. The idea behind this is that defining, updating and managing schemas should not be constrained to log types only. Since a full switch to SchemasAPI cannot happen until FE catches up, we have to retain the CustomLog endpoints along side the Schema endpoints. 

This PR implements the minimal endpoints required so that managed schemas can work properly throughout the system. Namely:
- GetSchema
- ListManagedSchemaUpdates
- UpdateManagedSchemas

kostaspap commented on internal/log_analysis/managedschemas/github.go:
```diff
@@ -0,0 +1,87 @@
+
+const githubAssetName = "managed-schemas.zip"
+
+func (p *GitHubRepository) ReleaseFeed(ctx context.Context, sinceTag string) ([]Release, error) {
```
In the end we followed a different approach compared to what @lindsey-w did, right? I am guessing we can look into merging the two codebases down the line.

alxarch replied:
The approach we took had to be functioning before the full PR on their side so it is bare-bones.
Once we have their code and have fully explored our requirements for both packs and managed schemas we can find some common way to handle this if needed.

kostaspap reviewed (commented):
Good stuff, nothing pops out in a first pass I did. 

Deferring approval until PR has a bit more information regarding testing.

kostaspap commented on internal/log_analysis/managedschemas/github.go:
```diff
@@ -0,0 +1,87 @@
+
+const githubAssetName = "managed-schemas.zip"
+
+func (p *GitHubRepository) ReleaseFeed(ctx context.Context, sinceTag string) ([]Release, error) {
```
Can you add a descriptions to the public methods?

alxarch replied:
Added

kostaspap commented on internal/log_analysis/managedschemas/managedschemas.go:
```diff
@@ -0,0 +1,154 @@
+	return nil
+}
+
+func ReadYAMLManifest(release string, r io.Reader) ([]ManifestEntry, error) {
```
Can you add a comment describing the structure on high level? Just enough to give the reader some context.

alxarch replied:
Added

alxarch commented on internal/core/logtypesapi/customlogs.go:
```diff
@@ -211,28 +201,21 @@ type DelCustomLogOutput struct {
-	for _, logType := range available {
-		if strings.HasPrefix(logType, customlogs.LogTypePrefix) {
-			custom = append(custom, logType)
+	records := make([]*SchemaRecord, 0, 16)
```
Just a non-zero pre-allocation. 
When using `append` if the capacity of a slice is not sufficient, `go` doubles* the capacity of the slice and copies data.
So the capacity of an empty slice goes:
`0, 1, 2, 4, 8, 16, 32,...`
By pre-allocating to `16` you can avoid the extra steps on the initial allocations.
This is obviously not needed here. Will remove, it is just a habit of initializing slices to some size from performance sensitive code.

*it actually sets it to `2*capacity+1`

alxarch commented on internal/log_analysis/datacatalog_updater/main/lambda.go:
```diff
@@ -83,17 +83,14 @@ func main() {
-			LogTypesAPI: logtypesAPI,
-		},
-	)
+	apiResolver := &logtypesapi.Resolver{
+		LogTypesAPI:    logtypesAPI,
+		NativeLogTypes: logtypes.MustMerge("native", registry.NativeLogTypes(), snapshotlogs.LogTypes()),
+	}
```
Yes, one of the core points this PR addresses

kostaspap reviewed (approved):
Code looks good.
Note that this should be tagged with `sync-to-release` tag.

alxarch commented:
Ran into some trouble while deploying for e2e tests.
I will fix any issues and re-request a review.

alxarch commented:
Fixed various issues with the exported YAML files.
Also fixed an issue brought up by @georgepsarakis that involved semver comparisons when updating a managed schema.

georgepsarakis commented on internal/log_analysis/managedschemas/build.sh:
```diff
@@ -0,0 +1,70 @@
+EOF
+
+# Clean up
+rm -rf "${TMP_DIR}"
```
I know that we create a temporary directory in a safely manner above, so it is unlikely for accidents to happen, but this seems a bit disconcerting. As an alternative you could do something like this:

```
BUILD_DIR="$TMPDIR/panther-managed-schemas-build/"$(date "+%Y%m%d%H%M%S")
mkdir -p "$BUILD_DIR"
...
rm -rf "$TMPDIR/panther-managed-schemas-build/*"
```

alxarch replied:
AFAIK `mktemp -d` is random each time. Also it allows the environment to determine where it wants its temporary directories to reside.
What kind of accidents are you referring to?

georgepsarakis replied:
It's better to depend on (some) constant parts rather than entirely on variables for catastrophic commands. I am not sure how this could happen with the current script, I just thought to point it out if it takes one more concern out of the equation, we can always revisit this later.

alxarch replied:
Seeing as this is to be run once when updating the minimum required version, I'll just remove the Clean up altogether

kostaspap commented on internal/core/logtypesapi/dynamodb.go:
```diff
@@ -33,89 +33,217 @@ import (
-	attrDeleted    = "IsDeleted"
-	attrRevision   = "revision"
-	attrLogType    = "logType"
+	attrRecordKind   = "RecordKind"
+	attrDisabled     = "IsDeleted"
+	attrRevision     = "revision"
+	attrDescription  = "description"
+	attrReferenceURL = "referenceURL"
+	attrSpec         = "logSpec"
+	attrManaged      = "managed"
+	attrLogType      = "logType"
```
Can you share some more context? What is happening here? Are we changing the way we store schemas in DDB?

alxarch replied:
We are now storing both managed and user-defined schemas. The difference between them is that managed schemas have `managed` field set to `true` and they have `release` set to their version tag. Also managed schemas keep no record of past revisions.

kostaspap commented on internal/core/logtypesapi/dynamodb.go:
```diff
@@ -33,89 +33,217 @@ import (
+		&transact.Update{
+			TableName: tableName,
+			Key:       schemaRecordKey(record.Name, 0),
+			Set: map[string]interface{}{
```
Please add comments there. What is the transaction logic here? Why do we have to use transactions here?

alxarch replied:
We use transactions to have coherent logic and error handling between endpoints. We could be using a simple `UpdateItem` for managed schemas but it would deviate from what we use for UserSchemas. Also building the update expression is much simpler using the `transact` package than doing it manually.

kostaspap commented on internal/core/logtypesapi/dynamodb.go:
```diff
@@ -33,89 +33,217 @@ import (
+						return e
+					}
+					if !rec.Managed {
+						return NewAPIError(ErrRevisionConflict, fmt.Sprintf("schema record %q is not managed", rec.RecordID))
```
Is `ErrRevisionConflict` the correct error code here?

alxarch replied:
Changed to AlreadyExists

kostaspap commented on internal/core/logtypesapi/dynamodb.go:
```diff
@@ -33,89 +33,217 @@ import (
+
+	output, err := d.DB.GetItemWithContext(ctx, &input)
+	if err != nil {
+		L(ctx).Error(`failed to get DynamoDB item`, zap.Error(err))
```
Shouldn't we log the error at the caller?

alxarch replied:
Removed the unused function

kostaspap commented on internal/core/logtypesapi/dynamodb.go:
```diff
@@ -33,89 +33,217 @@ import (
+		AvailableLogTypes []string
+	}{}
+	if err := dynamodbattribute.UnmarshalMap(output.Item, &item); err != nil {
+		L(ctx).Error(`failed to unmarshal DynamoDB item`, zap.Error(err))
```
Same as above. I think logging should be higher in the call stack?

alxarch replied:
Removed unused function

kostaspap commented on internal/core/logtypesapi/dynamodb.go:
```diff
@@ -156,77 +284,67 @@ func (d *DynamoDBLogTypes) DeleteCustomLog(ctx context.Context, id string, revis
 }
 
-func (d *DynamoDBLogTypes) CreateCustomLog(ctx context.Context, id string, params *CustomLog) (*CustomLogRecord, error) {
+func (d *DynamoDBSchemas) CreateUserSchema(ctx context.Context, id string, upd SchemaUpdate) (*SchemaRecord, error) {
```
Is `UserSchema` what we see in other places as `CustomSchema`? It would be better for readability purposes to stick to one terminology.

alxarch replied:
Yes we stick to one. `SchemaDatabase` has `Schema` actions.
The LogTypesAPI has to retain the `CustomLog` endpoints for compatibility.

kostaspap commented on internal/core/logtypesapi/dynamodb.go:
```diff
@@ -156,77 +284,67 @@ func (d *DynamoDBLogTypes) DeleteCustomLog(ctx context.Context, id string, revis
+	record := SchemaRecord{
+		Name:         id,
+		Revision:     1,
+		Managed:      false,
```
We use `Custom` and `Managed`. Instead of having `managed` true/false, should we just have `custom` true/false? or maybe we should have a `type` with value string that can take either `Managed` or `Custom` value? This might require bigger changes in this PR, would like to hear your thoughts.

alxarch replied:
See above, the terms are `UserSchema` and `ManagedSchema`. Since entries already in DDB are `UserSchemas` we have to use `Managed: true` as the boolean flag.

kostaspap commented on internal/core/logtypesapi/dynamodb.go:
```diff
@@ -254,31 +372,31 @@ func buildCreateRecordTx(tbl string, record CustomLogRecord) transact.Transactio
 }
 
-func (d *DynamoDBLogTypes) UpdateCustomLog(ctx context.Context, id string, revision int64, params *CustomLog) (*CustomLogRecord, error) {
+func (d *DynamoDBSchemas) UpdateUserSchema(ctx context.Context, id string, rev int64, upd SchemaUpdate) (*SchemaRecord, error) {
```
Why do we need separate Update functions but on `ListDeletedLogTypes ` function?

alxarch replied:
We need separate separate functions because `UserSchema` records are keeping historical records of previous revisions.
The `SchemaDatabase` interface is to be further simplified in an upcoming PR.
I don't understand the second part of the question after 'but'.

kostaspap replied:
I meant to write "Why we don't have different `ListDeleted` functions for User/Managed schemas". It is clear to me now though.

kostaspap commented on internal/core/logtypesapi/dynamodb.go:
```diff
@@ -400,10 +515,10 @@ func newStringSet(strings ...string) *dynamodb.AttributeValue {
 }
 
-func (d *DynamoDBLogTypes) ListDeletedLogTypes(ctx context.Context) ([]string, error) {
+func (d *DynamoDBSchemas) ListDeletedLogTypes(ctx context.Context) ([]string, error) {
```
nit: renamed to `ListDisabledSchemas`?

alxarch replied:
I'll remove this, it's unused

kostaspap commented on internal/core/logtypesapi/inmem.go:
```diff
@@ -21,46 +21,39 @@ package logtypesapi
 import (
 	"context"
 	"fmt"
+	"strings"
```
Fine to keep this change here but a more general thought: Having to maintain a in-memory version of the schema's repo will be challenge to maintain long term as it requires its own tests to verify it is actually working as expected etc. A mocking library could be used for unit tests that don't want to interact with DDB.

alxarch replied:
Yeah, the logic was to keep inmem for tests but that is also cumbersome. I should just add a mock for the `SchemasDatabase` interface and use that in tests.

alxarch replied:
The `SchemaDatabase` interface will be simplified in a subsequent PR. This will trim down on the code we require to maintain for `InMem`

kostaspap commented on internal/core/logtypesapi/managed_schemas.go:
```diff
@@ -0,0 +1,148 @@
+}
+
+func loadManifest(ctx context.Context, release, manifestURL string) ([]managedschemas.ManifestEntry, error) {
+	if release == managedschemas.ReleaseVersion {
+		return managedschemas.LoadDefaultManifest()
+	}
+	return managedschemas.LoadReleaseManifestFromURL(ctx, manifestURL)
```
Why is this check needed? can you add a comment?

alxarch replied:
We want to avoid having to call github on deployment

kostaspap commented on internal/core/logtypesapi/resolver.go:
```diff
@@ -31,23 +30,21 @@ import (
-		LogType:  name,
+	reply, err := r.LogTypesAPI.GetSchema(ctx, &GetSchemaInput{
+		Name:     name,
 		Revision: 0,
```
Is revision 0 meant to be the latest revision?

alxarch replied:
Yes

kostaspap commented on internal/log_analysis/log_processor/logschema/infer.go:
```diff
@@ -247,6 +247,12 @@ func inferFieldSchema(s *FieldSchema, field reflect.StructField) error {
 		value.IsEventTime = isEventTime
 	}
 	value.TimeFormat = field.Tag.Get("tcodec")
+	if value.Type == TypeTimestamp {
+		value.TimeFormat = "rfc3339"
+		if format, ok := field.Tag.Lookup("tcodec"); ok {
+			value.TimeFormat = format
+		}
+	}
```
Why is this part of this PR? What is this change for?

alxarch replied:
The exported YAML files were not working properly.

kostaspap reviewed (commented):
Woah, big PR! Put some comments, nothing seems like a blocker, mostly changes related to readability

alxarch commented on deployments/log_analysis.yml:
```diff
@@ -115,9 +115,17 @@ Conditions:
   ###### Update Glue Table Schemas for Deployed Tables #####
   UpdateLogProcessorGlueTables:
-    DependsOn: UpdaterFunction
+    DependsOn: UpdateManagedSchemas
```
```suggestion
    DependsOn: UpdateManagedSchemas # The data catalog updater needs to have up-to-date schemas 
```

alxarch commented on internal/core/logtypesapi/api.go:
```diff
@@ -38,31 +39,36 @@ const LambdaName = "panther-logtypes-api"
+	// UpdateManagedSchema updates a managed schema to the release version provided
+	UpdateManagedSchema(ctx context.Context, id string, release string, upd SchemaUpdate) (*SchemaRecord, error)
+
+	ToggleSchema(ctx context.Context, id string, enabled bool) error
```
added

alxarch commented on internal/core/logtypesapi/customlogs.go:
```diff
@@ -136,31 +130,27 @@ func (api *LogTypesAPI) checkUpdate(a, b *logschema.Schema) error {
+	}
+	if refURL != "" {
+		schema.ReferenceURL = refURL
+	}
```
What this does is to use the values in the form fields as _overrides_ for the values in the YAML.
This is the behavior we want here. Having form fields that overlap with the values in the YAML is not ideal but it allows the user to focus on defining the fields in YAML.
We also do not want to be modifying the user's YAML _input_ so the only point where we can apply these overrides is in `buildSchema`, _before_ validating the user input.  OTOH we are reusing this function to handle managed schema updates where the overrides would be empty.
I updated the code to distinguish between `buildAndValidateUserSchema` and `buildAndValidateManagedSchema`.
